### PR TITLE
New container for connections

### PR DESCRIPTION
**Pull request by hakonsbm**
With 5g, connections are all put in one big vector. As the number of connections becomes large, memory consumption from vector growth becomes increasingly problematic. Some attempts were made to limit this problem (see jakobj/nest-simulator#116) and we also tried using std::deque as a container, both with limited success. Memory consumption is still bad, and random access with std::deque is inefficient and therefore slowed down simulations.

This PR implements a specialized container, named BlockVector, holding data in a vector of vectors. The size of each block (inner vector) is fixed to a number that is a power of 2, allowing use of bitwise operators to map an index to the right block and position in that block. Because of its fragmented data-structure, memory is better managed than when using a single vector, and using bitwise operators makes random access close to as efficient as using vector.

Additionally, when sorting connections, the `exchange_()` function is replaced with `std::swap()`, which somewhat reduces sorting time.

### CODE REVIEW

**Comment by ikitayama** [comment]
@hakonsbm - do you have some performance data collected on a supercomputer?

**Comment by heplesser** [comment]
Some benchmark results attached, including from Jureca.
[Bench_214_sq_db70dcc1edb-2.pdf](https://github.com/nest/nest-simulator/files/2505835/Bench_214_sq_db70dcc1edb-2.pdf)

File: libnestutil/block_vector.h [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+}
+
+template < typename value_type_ >
+inline typename BlockVector< value_type_ >::iterator
+BlockVector< value_type_ >::erase( const_iterator first, const_iterator last )
+{
+  assert( first.block_vector_ == this );
+  assert( last.block_vector_ == this );
+  assert( last < finish_ or last == finish_ );
+  // TODO: Except for some special cases, this method may kill performance as
```
Comment (jakobj) [inline libnestutil/block_vector.h]: should these todos be addressed before merging?

File: nestkernel/source_table_position.h [hunk 2]
```diff
@@ -46,6 +48,9 @@ struct SourceTablePosition
   template < typename T >
```
Comment (jakobj) [inline nestkernel/source_table_position.h]: remove outdated function

File: nestkernel/connector_base.h [hunk 3]
```diff
@@ -174,6 +174,7 @@ class ConnectorBase
    * Sort connections according to source gids.
    */
   virtual void sort_connections( std::vector< Source >& ) = 0;
```
Comment (jakobj) [inline nestkernel/connector_base.h]: remove outdated function

File: nestkernel/connector_base.h [hunk 4]
```diff
@@ -451,7 +450,7 @@ class Connector : public ConnectorBase
   void
   reserve( const size_t count )
   {
-    C_.reserve( count );
+    //    C_.reserve( count );
```
Comment (jakobj) [inline nestkernel/connector_base.h]: maybe add a comment why this is commented out, or remove function entirely if not used?

File: libnestutil/sort.h [hunk 5]
```diff
@@ -180,7 +282,14 @@ quicksort3way( std::vector< T1 >& vec_sort,
  */
 template < typename T1, typename T2 >
 void
-sort( std::vector< T1 >& vec_sort, std::vector< T2 >& vec_perm )
+sort( std::vector< T1 >& vec_sort, BlockVector< T2 >& vec_perm )
```
Comment (jakobj) [inline libnestutil/sort.h]: necessary? (see above)

File: libnestutil/sort.h [hunk 6]
```diff
@@ -68,7 +70,7 @@ median3_( const std::vector< T >& vec,
 template < typename T1, typename T2 >
 void
 insertion_sort( std::vector< T1 >& vec_sort,
```
Comment (jakobj) [inline libnestutil/sort.h]: why do we need the combination of `vector` and `BlockVector`? shouldn't the sorting always be called on two `BlockVector`s? there's heavy code duplication here, so either this function should be removed or integrated with the sort on two `BlockVector`s

File: testsuite/cpptests/test_sort.h [hunk 7]
```diff
@@ -49,51 +49,102 @@ is_sorted( std::vector< size_t >::const_iterator begin,
...
 
 /**
  * Tests whether two arrays with randomly generated numbers are sorted
  * correctly by a single call to sort.
  */
-BOOST_AUTO_TEST_CASE( test_random )
+BOOST_AUTO_TEST_CASE( test_random_vec_seq )
 {
   const size_t N = 20000;
   std::vector< size_t > vec0( N );
```
Comment (jakobj) [inline testsuite/cpptests/test_sort.h]: why test the combination of `vector` and `BlockVector`?

File: libnestutil/block_vector.h (hunk 1 above)
Comment (hakonsbm) [reply libnestutil/block_vector.h]: The function isn't that much used anyway, and making it more efficient would require quite a bit of mental work. So I removed the TODO.

File: nestkernel/source_table_position.h (hunk 2 above)
Comment (hakonsbm) [reply nestkernel/source_table_position.h]: Removed!

File: nestkernel/connector_base.h (hunk 3 above)
Comment (hakonsbm) [reply nestkernel/connector_base.h]: Removed!

File: nestkernel/connector_base.h (hunk 4 above)
Comment (hakonsbm) [reply nestkernel/connector_base.h]: I removed the function entirely, as one can't (and shouldn't) reserve memory for the new container.

File: libnestutil/sort.h (hunk 5 above)
Comment (hakonsbm) [reply libnestutil/sort.h]: Removed it!

**Comment by hakonsbm** [comment]
@jakobj Thanks for your review! I have updated the PR and addressed your comments.

File: libnestutil/sort.h [hunk 8]
```diff
@@ -23,32 +23,34 @@
...
 /**
- * Exchanges elements i and j in vector vec.
+ * Calculates the median of three elements.
+ * See http://algs4.cs.princeton.edu/23quicksort/QuickX.java.html.
  */
 template < typename T >
-inline void
-exchange_( std::vector< T >& vec, const size_t i, const size_t j )
+inline size_t
+median3_( const std::vector< T >& vec,
```
Comment (jakobj) [inline libnestutil/sort.h]: this can probably also be removed

**Comment by hakonsbm** [comment]
@jakobj I agree, `reserve_connections` in `GenericConnectorModel` can be removed. And how about [this one](https://github.com/hakonsbm/nest-simulator/blob/098fc61b2a6bc3e46206fda2ee246ee58402b3bc/nestkernel/connection_manager.cpp#L1453) in `connection_manager`?

**Comment by jakobj** [comment]
I would remove those and the corresponding calls from `conn_builder.cpp` as well. Otherwise we will wonder soon why these function exist ;)

**Comment by heplesser** [comment]
@hakonsbm I agree with the removal of the `reserve()` calls for now. But could you create a branch or tag (at least in your repo) that preserves them---maybe we will find a way of using them for optimization in the future.

**Comment by ikitayama** [comment]
@hakonsbm I'd like to test the latest development of your branch if possible.

**Comment by hakonsbm** [comment]
@heplesser I have created a [tag in my repo](https://github.com/hakonsbm/nest-simulator/releases/tag/before_removing_reserve).

@ikitayama Please do! :+1: 
https://github.com/hakonsbm/nest-simulator/tree/new_container

**Comment by ikitayama** [comment]
I’ll share the results with you and the usual suspects off the list alright?

**Comment by hakonsbm** [comment]
That's okay for me.

**Pull request merged**